src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -904,32 +919,29 @@ private unsafe void SetupOverlappedSingle(bool pinSingleBuffer)
         }
 
         // Sets up an Overlapped object with multiple buffers pinned.
-        private unsafe void SetupOverlappedMultiple()
+        private unsafe void SetupPinHandlesMultiple()
         {
             int bufferCount = _bufferListInternal.Count;
 
             // Number of things to pin is number of buffers.
             // Ensure we have properly sized object array.
-            if (_objectsToPin == null || (_objectsToPin.Length != bufferCount))
+            if (_multipleBufferGCHandles == null || (_multipleBufferGCHandles.Length < bufferCount))
             {
-                _objectsToPin = new object[bufferCount];
+                _multipleBufferGCHandles = new GCHandle[bufferCount];
```

Comment:
Before we overwrite arrays like this, consider asserting that none of the handles in it are allocated/valid.